### PR TITLE
GEODE-6011: Prevent synchronization using strings and boolean

### PR DESCRIPTION
**Pull request by nabarunnag**
* Locking is now achieved using reentrant locks as they are more fair than synchronization.

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

**Comment by pivotal-jbarrett** [comment]
@nabarunnag any thoughts on using an autoclosable wrapper and using with resource to guard these locks?

**Comment by nabarunnag** [comment]
@pivotal-jbarrett Thank you for your suggestion and I felt that it was a good idea to get the lock under a TWR.
But after going through all the articles and Java's Project Coin's decision to not include the locks as TWR, on which was creating the instance every time something needs to be locked, I am planning to use pure reentrant locks for this PR. Maybe, if needed, we can create a new JIRA to implement this feature in a broader scope of the project rather than just two files.

**Comment by pivotal-jbarrett** [comment]
@nabarunnag Good point on the object creation overhead. Been in C++ land so long I didn't even think about that cost of RAII in Java.

File: extensions/geode-modules/src/main/java/org/apache/geode/modules/util/BootstrappingFunction.java [hunk 1]
```diff
@@ -42,6 +43,7 @@
   private static final long serialVersionUID = 1856043174458190605L;
 
   public static final String ID = "bootstrapping-function";
+  private static final ReentrantLock registerFuntionLock = new ReentrantLock();
```
Comment (galen-pivotal) [inline extensions/geode-modules/src/main/java/org/apache/geode/modules/util/BootstrappingFunction.java]: typo: should be `registerFunctionLock`

**Pull request merged**